crates/sparrow-runtime/src/prepare/prepare_input_stream.rs

```diff
@@ -0,0 +1,144 @@
+use anyhow::Context;
+use arrow::compute::SortColumn;
+use arrow::datatypes::{ArrowPrimitiveType, TimestampNanosecondType};
+use arrow::error::ArrowError;
+use arrow::record_batch::RecordBatch;
+use error_stack::{IntoReport, ResultExt};
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use itertools::Itertools;
+use sparrow_api::kaskada::v1alpha::{slice_plan, TableConfig};
+use sparrow_core::TableSchema;
+
+use crate::prepare::slice_preparer::SlicePreparer;
+use crate::prepare::Error;
+use crate::RawMetadata;
+
+use super::column_behavior::ColumnBehavior;
+use super::PrepareMetadata;
+
+/// An stream over prepare batches and corresponding key hash metadata.
+///
+/// This stream reads unordered, unprepared batches from the `reader`, and is
+/// responsible for:
+/// 1. Inserting the time column, subsort column, and key hash from source to
+///    the batch
+/// 2. Casting required columns
+/// 3. Sorting the record batches by the time column, subsort column, and key hash
+/// 4. Computing the key-hash and key batch metadata.
+pub async fn try_new<'a>(
```

Comment:
Would suggest calling this something different -- like `prepare_input(...)`. Basically:

1. We no longer need a `struct` so there isn't a type, so `try_new` is less appropriate.
2. The type reflects this is a stream.

Then, we can write `crate::prepare::prepare_input(reader_stream, ...)` which would be pretty clear.